### ISSUE TITLE
Phone number formatting

### ISSUE THREAD

**Issue by bobintetley** [bobintetley/asm3#499]
Just like we have time and date widgets, have a phone number widget. It should perform some basic formatting of values inserted into it. Values are checked on set and blur.

1. If locale == "en" and we have a 10 digit number, format to (ddd) ddd-dddd
2. If locale == "en_GB" and we have an 11 digit number, format to (ddddd) dddddd

That's it for now, do nothing for anything else and treat as free text like it's always been.

**Comment by julian-poidevin** [comment]
Hi @bobintetley, I'll try to work on that as it is a "good first issue".

I already changed `i18n.py` to add functions to define phone format per local and functions to format the phone : 

```python
def format_phone(local, p):
    input_regex = get_locale_map(locale, LM_PHONE_INPUT_FORMAT_REGEX)
    output_regex = get_locale_map(locale, LM_PHONE_OUTPUT_FORMAT_REGEX)
    return re.sub("r" + input_regex, "r" + output_regex, p)
```

I also added fields in the `locale_maps` array to set the `LM_PHONE_INPUT_FORMAT_REGEX `and `LM_PHONE_OUTPUT_FORMAT_REGEX `fields.

I'll now like to change the way the phones are displayed in the person header, I'm trying to change the `header_edit_header.js` file and particularly `person_edit_header: function(p, selected, counts)` function in order to display a formatted phone number.

I have trouble understanding how the js code interacts with python. As I understand, the following line should be changed : 

https://github.com/bobintetley/asm3/blob/c53d209da117c3e0e62fc5968b7b7ea6d8f334ba/src/static/js/header_edit_header.js#L449

In which js file should I add a function to call the python `format_phone` function ?

Thanks

**Comment by bobintetley** [comment]
None of the phone number formatting should be done in the Python code and there shouldn't be any need to change header_edit_header.js

Instead, you should pass the phone format to the JS as part of the return value from code.py/configjs - you can then access it through the "asm" global in the JS frontend.

Then, you should have a text widget for phone numbers (similar to the ones we have already for numbers, ip addresses, etc). These live in common_widgets.js

All your formatting should be handled in your new js widget after the user leaves the phone number widget (blur event) rather than in the python code, since the js is the presentation layer. 

Give your new widget a class (eg: asm-phonenumberbox) and then respond to that class and create the widget in common.js/bind_widgets as we do for others.

There's no harm in storing the formatted phone numbers in the database because the search code strips the non-numeric portion when comparing and it means that what header_edit_header.js displays in the banner will be the formatted value too.

So, to summarise - phone number format in the locale in i18n, return it in code.py/configjs, phone number text widget it in common_widgets.js, update person_new.js, common_personchooser.js and person.js to use the new phone widget by adding your asm-phonenumberbox class to their input tag instead of asm-textbox.
